python/ml4ir/applications/ranking/model/metrics/helpers/metrics_helper.py

```diff
@@ -16,9 +17,83 @@ class RankingConstants:
+    Adds a new column indicating the top graded relevance record per query in the DataFrame.
+
+    Args:
+        df (pandas.DataFrame): The input DataFrame.
+        label_col (str): The column name containing the relevance scores.
+        new_col_name (str): The name for the new column indicating the top graded relevance.
+
+    Returns:
+        pandas.DataFrame: The input DataFrame with the new column added.
+        """
+    # remove queries with no relevance scores
+    grouped_sum = df.groupby('query_id')[label_col].sum()
+    queries_with_relevance_scores = grouped_sum[grouped_sum > 0]
+    df_with_relevance_scores = df.loc[df["query_id"].isin(queries_with_relevance_scores.index)]
+
+    # Group the DataFrame by 'query_id' and find the record with the highest 'target_relevance'
+    top_records = df_with_relevance_scores.groupby('query_id')[label_col].idxmax()
+
+    # Create a new column 'top_target_relevance' and initialize with 0
+    df_with_relevance_scores[new_col_name] = 0.0
+
+    # Set the 'top_target_relevance' value to 1 for the top records
+    df_with_relevance_scores.loc[top_records.values, new_col_name] = 1.0
+
+    return df_with_relevance_scores
+
+
+def compute_ndcg(df, label_col, pred_col="ranking_score", new_col="new_NDCG"):
+    """
+    Computes the Normalized Discounted Cumulative Gain (NDCG) for each query in the DataFrame.
+
+    Args:
+        df (pandas.DataFrame): The input DataFrame.
+        label_col (str): The column name containing the true relevance scores.
+        pred_col (str, optional): The column name containing the predicted scores. Default is "ranking_score".
+        new_col (str, optional): The name for the new column containing the computed NDCG values. Default is "new_NDCG".
+
+    Returns:
+        pandas.DataFrame: The input DataFrame with the new column containing the computed NDCG values.
+    """
+    grouped = df.groupby('query_id')
+    for qid, group in grouped:
+
+        y_true = group[label_col].values
+        y_pred = group[pred_col].values
+
+        # Sort the predicted scores in descending order
+        sorted_indices = np.argsort(-y_pred, axis=-1)
+        sorted_labels = np.take_along_axis(y_true, sorted_indices, axis=-1)
+
+        # Compute the Discounted Cumulative Gain (DCG)
+        positions = np.arange(1, y_true.shape[-1] + 1, dtype=float)
+        discounts = np.log1p(positions) / np.log1p(2.0)  # log base 2
+        dcg = np.sum(sorted_labels / discounts, axis=-1)
+
+        # Sort the true relevance scores in descending order
+        true_sorted_indices = np.argsort(-y_true, axis=-1)
+        true_sorted_labels = np.take_along_axis(y_true, true_sorted_indices, axis=-1)
+
+        # Compute the Ideal Discounted Cumulative Gain (IDCG)
+        idcg = np.sum(true_sorted_labels / discounts, axis=-1)
+
+        # Compute the Normalized Discounted Cumulative Gain (NDCG)
+        ndcg = dcg / idcg
+
+        df.at[df.loc[df["query_id"] == qid].index, new_col] = ndcg
```

Comment:
@mohazahran Can you use the same function we use for RankMF? I see a duplication otherwise.

Reply:
ping

Reply:
I thought I replied to this comment. It was not trivial to do and would require changes. I don't think this should be blocking this PR though